[서버 사이드 렌더링 - 1단계] 해시(김가연) 미션 제출합니다.

dle234 opened the pull request:
//TODO: 크론이 주신 코드 반영을 아직 안했습니다🥹🥹🥹🥹🥹🥹🥹//

안녕하세요 해리 👋
오랜만이네요~~~~ 리뷰 잘 부탁해욥

## 👀 리뷰를 통한 생각 나눔

SSR 렌더링 시 초기 렌더링 성능이 왜 유리할까?

## 1. 
### CSR
<img width="494" alt="image" src="https://github.com/user-attachments/assets/b480a559-d76e-4332-b2dd-e80ce65d1ace">

url -> 웹서버에서 기본적인 html, js , css 정적 파일을 보내줌 -> 브라우저에서 렌더링 -> 브라우저에서 fetch 해서 서버에 데이터 요청 -> 받아온 데이터를 받아 브라우저에서 그 부분을 갈아끼워줌

### SSR
<img width="590" alt="image" src="https://github.com/user-attachments/assets/9102b71c-13cc-466e-80bf-dbd65e2280e0">

url -> 경로에 맞는 정적 파일을 웹서버에 요청 -> SSR 서버에서(node js) 리액트 컴포넌트를 렌더링 , 그 과정에서 영화 데이터를 fetch 를 해서 넣어주고 html 페이지 생성. -> 완성된 html 파일 브라우저에 보내줌 -> 브라우저에서 html 렌더링 

즉 SSR 은 리액트를 서버에서 렌더링 해주고, 만들어진 html 파일만 브라우저에서 렌더링 해주고,
CSR 은 html + css + js 파일을 모두 렌더링 후 브라우저 환경에서 번들링 및 트랜스파일링 된 파일을 돌려준다. 그렇게 만들어진 html 파일을 렌더링 한다. 
그래서 SSR이 초기 렌더링이 더 빠르다.

❓궁금한점❓ 
SSR 을 위해 구현한 node js 코드는 정적 배포(s3) 가 안될 것 같은데, 그러면 이 코드를 위해 서버를 새로 구축해야될까요..? 그리고 SSR 로만 구현하게 된다면 정적 배포가 필요하지 않을까요?(서버에서 html 을 불러오고 수정 후 브라우저에 보내주니까..)
여기에 대한 리뷰어님의 생각이 궁금해요!

## 2. 
서버 측에서 클라이언트 측 컴포넌트 코드를 불러올 때 ReactDOM.hydrateRoot() 함수를 사용하고 있다.
일단 초기에 html 파일을 렌더링 하고 이후 동적으로 상호작용 해야 하는 부분이 있다면 CSR 을 활용하기 위해 리액트 앱을 불러와야 한다. 
이때, 초기의 html 에서 동적인 이벤트를 붙이는 과정을 하이드레이션 이라고 하는데, 하이드레이션을 하기 위해 리액트 앱에서의 html 과 서버에서 생성된 Html 파일이 일치해야 한다. 그러기 위해 window.__INITIAL_DATA__ 를 활용해 전역 객체를 초기화 해 주어야 하는 것 같다.


## 3. 아래와 같은 오류가 왜 발생하는지 탐구해 보세요.
```js
react-dom.development.js:86 Warning: An error occurred during hydration. The server HTML was replaced with client content in <div>
react-dom.development.js:12507 Uncaught Error: Hydration failed because the initial UI does not match what was rendered on the server.
```
React를 포함한 JavaScript 파일 로드, 실행 -> 하이드레이션 시작 -> DOM을 순회하며 서버에서 렌더링된 HTML과 클라이언트에서 렌더링하려는 가상 DOM 비교 -> 기존 HTML에 이벤트 리스너를 연결 -> 상호작용 가능한 동적 웹 완성

하이드레이션을 하는 도중 초기에 렌더링한 html과 리액트 컴포넌트의 내용이 달라서 에러가 나는 것 같습니다. 

❓궁금한점❓ 
1. 하이드레이션 할 때 리액트 App이 만든 가상돔과 브라우저에 초기렌더링 된 html 을 비교해야할텐데, 비교할때 돔이 같아야 하잖아요,, 그렇다면 html 의 내용들을 App 으로 모두 구현해야 하는걸까요? 그러면 컴포넌트에서 분기처리 한 것도 구현되어야 하고.. 고려할 것도 많아서 너무 번거로운 것 같아요 🥹 (아니면 비교하는 로직이 다를까요? )

2. 그리고 하이드레이션의 타이밍도 궁금합니다. 지금은 hydration이 필요 없어서인지(동적인 상호작용이 일어나지 않아서 그런지) 없어도 잘 동작하는데, 언제 하이드레이션이 일어나게 되는지 잘 이해가 가지 않았어요.
브라우저에서 첫 페이지를 렌더링 후 어떤 코드가 하이드레이션 하는 코드를 트리거 하는 것 일까요??

 여기에 대한 리뷰어님의 생각이 궁금해요!

dle234 committed:
feat: movies 불러오기

dle234 committed:
feat: Header 컴포넌트 구현

dle234 committed:
feat: 전역 초기화 객체 추가

dle234 committed:
refactor: movies -> initialData 로 수정

dle234 committed:
feat: App 컴포넌트 구성

Commit:
fix: 뼈대 코드 수정 및 샘플 코드 제공

hwinkr commented on src/apis/fetchMovies.js:
```diff
@@ -0,0 +1,9 @@
+
+  const data = await response.json();
+
+  return data.results || [];
```
결과가 존재하지 않을 경우, 빈 배열을 반환하도록 하는 예외 처리 좋네요 👏

hwinkr commented on src/server/routes/index.js:
```diff
@@ -6,30 +6,48 @@ import { fileURLToPath } from "url";
+    res.send(renderedHTML);
+  } catch (error) {
+    console.error("Error during rendering:", error);
+    res.status(500).send("An error occurred while rendering the page");
```
여기도 예외처리 굿 👍

hwinkr commented on src/server/routes/index.js:
```diff
@@ -6,30 +6,48 @@ import { fileURLToPath } from "url";
+    const template = fs.readFileSync(templatePath, "utf-8");
+    const movies = await fetchNowPlayingMovieItems();
+    const renderedApp = renderToString(<App movies={movies} />);
+    const renderedHeader = renderToString(<Header movie={movies[0]} />);
```
단순 궁금증인데, App과 Header를 따로 문자열로 변환해 주신 이유가 있을까요? App을 문자열로 반환하면 그 안에 Header가 포함되지 않을까 싶어서요!

dle234 replied:
index.html 을 최대한 변경 없이 구현하려고 그렇게 부분적으로 replace 해주었는데, 생각해보니 App 자체를 바꿔줘야 하이드레이션 할 때 싱크가 잘 맞을 것 같네요🥲 수정하겠습니다~~~

hwinkr commented on src/client/components/Home.jsx:
```diff
@@ -1,7 +1,26 @@
+          ))}
+        </ul>
+      ) : (
+        <p>영화가 없어요</p>
```
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ fallback UI 좋네요

hwinkr commented on src/server/routes/index.js:
```diff
@@ -22,18 +22,18 @@ router.get("/", async (_, res) => {
+    `
+    );
+
+    const headerHtml = initData.replace(
```
headerHtml이 아래에 renderedHTML과 변수명이 약간 달라서 통일되었으면 하는데 해시의 생각은 어떤가요?

그리고, 아래에 renderedHTML에서 renderedApp으로 대체하고 있는데 headerHtml에서 renderedHeader으로 대체하고 있어서 변수명을 통일하면 좋을 것 같아요!

headerHTML, AppHTML이런 식으로요

dle234 replied:
오 이 부분은 좀 헷갈릴 수 있을 것 같네요 ! 감사합니다~~~👍

hwinkr commented on src/server/routes/index.js:
```diff
@@ -19,8 +19,8 @@ router.get("/", async (_, res) => {
     const template = fs.readFileSync(templatePath, "utf-8");
-    const movies = await fetchNowPlayingMovieItems();
-    const renderedApp = renderToString(<App movies={movies} />);
+    const movies = (await fetchNowPlayingMovieItems()) || [];
```
👏

hwinkr commented on src/client/components/Footer.jsx:
```diff
@@ -0,0 +1,12 @@
+    <footer className="footer">
+      <p>&copy; 우아한테크코스 All Rights Reserved.</p>
+      <p>
+        <img src="" width="180" />
```
여기 이미지가 없는 것 같아요!

dle234 replied:
헉 🥹 이미지 넣는걸 놓쳤네요 ,,

hwinkr reviewed (changes requested):
안녕하세요 해시~ 빠르게 잘 구현해 주셨네요! 저는 미션을 늦게 시작해서 변경된 코드로 진행해서 코드가 좀 다르네요~ 그래도 리뷰하는데 병목은 되지 않았습니다!

크게 수정해주실 부분은 없는 것 같고, 궁금한 점과 이미지가 빠진 것 같아서 이 부분만 해결해 주시면 될 것 같아요!! 다시 리뷰 요청 주세요~ #️⃣

질문으로 남겨주신 사항들은 저도 학습하고 있는 내용이라 충분히 학습하고 나서 같이 고민해보죠!

dle234 committed:
fix: 뼈대 코드 Pull 및 충돌 해결

dle234 committed:
fix: 수정된 템플릿 반영 및 구조 수정
- index.html 에 id root 만 넣어서 서버에서 제공하는 html 과 리액트앱이 일치하도록 수정
- Footer, Header 에 이미지 import
- 하이드레이션 확인 코드 추가

dle234 committed:
Merge pull request #1 from dle234/test

Test

dle234 committed:
feat: server 에 movies fetch api  로직 추가

dle234 committed:
refactor: replace 로직 분리

dle234 reviewed (commented):
해리 꼼꼼한 리뷰 감사해요~~~ 리뷰 반영하고 바뀐 뼈대 코드도 넣었어요!

수정사항
- 리뷰 반영
- 뼈대 코드 반영
  - process.env 를 브라우저 환경에서 읽으려면 번들링 과정을 바꿔야 하는 것 같아서 서버에서만 데이터를 불러올 수 있게 api/movies 경로를 추가했습니다.
  - 하이드레이션이 잘 됐는지 확인하기 위해 MovieItem 에 이벤트를 넣어주었습니다.

hwinkr reviewed (approved):
해시~ 리뷰 반영해주신 것 잘 확인했습니다. 해시의 질문들을 통해서 저도 학습할 수 있었어요. 해시의 질문에 대한 제 생각을 답변 드릴게요.

> *SSR 을 위해 구현한 node js 코드는 정적 배포(s3) 가 안될 것 같은데, 그러면 이 코드를 위해 서버를 새로구축해야될까요..? 그리고 SSR 로만 구현하게 된다면 정적 배포가 필요하지 않을까요?(서버에서 html 을 불러오고 수정 후 브라우저에 보내주니까..)
여기에 대한 리뷰어님의 생각이 궁금해요!*
> 

*SSR 을 위해 구현한 node js 코드는 정적 배포(s3) 가 안될 것 같은데,*

→ 저도 해시의 의견에 공감합니다! 현재 레벨3, 4에서 리액트로 만든 프로젝트를 배포하기 위해서 aws S3 버킷을 사용하고 있는데요. 해당 버킷 내부에는 html, css, js, img들과 같은 정적 파일들이 있고 서비스에 접근한 사용자는 버킷에 있는 정적 파일들을 다운로드 받게 됩니다. 하지만, SSR 프로젝트는 조금 얘기가 다를 것 같아요. 미션을 기준으로 제 생각을 전달해 보자면, Node.js 서버가 TMDB에서 영화 정보를 가져오기 위한 API 요청을 하고 정보를 받아온 후 HTML을 구성해서 다시 클라이언트로 전달해주게 됩니다. 그래서 SSR 프로젝트는 정적 파일들을 배포하는 것이 아니라, Node.js 서버를 배포한다고 생각하기 때문에 정적 파일들을 보관하는 S3로는 배포가 안될 것 같다는 것이 저의 생각입니다. 

*그리고 SSR 로만 구현하게 된다면 정적 배포가 필요하지 않을까요?*

→ SSR 프로젝트더라도 하이드레이션 과정을 거친다면 bundle.js와 같은 정적 파일들을 다운로드 받아야 할텐데요. 그래서 정적 배포가 필요하다고도 생각합니다.

결론적으로 SSR 프로젝트를 배포하고, aws 호스팅 서비스를 사용한다면 EC2 인스턴스를 사용해서 Node.js 서버와 bundle.js와 같은 정적 파일들을 배포하지 않을까…하는 것이 저의 의견입니다.

> *하이드레이션 할 때 리액트 App이 만든 가상돔과 브라우저에 초기렌더링 된 html 을 비교해야할텐데, 비교할때 돔이 같아야 하잖아요,, 그렇다면 html 의 내용들을 App 으로 모두 구현해야 하는걸까요? 그러면 컴포넌트에서 분기처리 한 것도 구현되어야 하고.. 고려할 것도 많아서 너무 번거로운 것 같아요 🥹 (아니면 비교하는 로직이 다를까요? )*
> 

모두 App으로 구현하는 것은 아니지 않을까요? 왜냐하면 이번 미션의 폴더 구조를 봤을 때 `client` 폴더 내부에는 index.html이 `views` 폴더 내부에 있는데요. App은 `document.getElementById("root")` 내부 DOM 내용만 채우고 있는 것 같네요. 그래서 비교를 할 때도 client 폴더 내부의 App 컴포넌트의 내용과 server 폴더 내부의 renderToString(App)을 한 후, 채워진 내용을 비교할 것 같아요. 

초기에 서버에서 보내주는 HTML, 클라이언트의 App 컴포넌트를 비교하기 위해서 App 컴포넌트 내부에서 분기처리하는 로직을 아예 없애거나 최소한으로 하지 않을까…생각합니다.

> *그리고 하이드레이션의 타이밍도 궁금합니다. 지금은 hydration이 필요 없어서인지(동적인 상호작용이 일어나지 않아서 그런지) 없어도 잘 동작하는데, 언제 하이드레이션이 일어나게 되는지 잘 이해가 가지 않았어요.*
> 
> 
> *브라우저에서 첫 페이지를 렌더링 후 어떤 코드가 하이드레이션 하는 코드를 트리거 하는 것 일까요??*
>

<img width="513" alt="image" src="https://github.com/user-attachments/assets/83be68b1-a72e-4610-997d-82be05cdd34f">

같이 순서에 대해서 생각해 봅시다. 이번 미션을 브라우저에서 실행해 보면 이미지처럼 <script /> 태그에 번들링되 자바스크립트를 다운로드 받는 것을 확인할 수 있네요. 

우선, 하이드레이션(hydration)은 서버에서 렌더링된 HTML과 클라이언트에서 React가 렌더링하는 Virtual DOM을 결합하여 React가 해당 DOM을 제어할 수 있게 하는 과정입니다. 이 과정을 거치고 나면 사용자와 상호작용이 가능해지겠죠. `hydrateRoot` 함수를 호출해서, 하이드레이션되는 과정은 다음과 같이 정리해볼 수 있을 것 같아요.하이드레이션은 클라이언트에서 `hydrateRoot` 함수가 호출될 때 발생합니다. 구체적인 타이밍을 단계별로 설명하자면

- 서버에서 `renderToString`을 사용하여 `App` 컴포넌트의 HTML을 생성하고, 이를 클라이언트에 전달
- → 이때 HTML 내에는 `id="root"`가 있는 `<div>` 태그와 서버에서 렌더링된 정적 콘텐츠가 포함
- → 브라우저는 서버로부터 전달받은 HTML을 파싱하고, 사용자는 초기 화면 확인
- → 이 시점에는 아직 React와 같은 JavaScript 로직이 실행되지 않았으므로 동적인 상호작용은 불가. 즉, 정적인 컨텐츠에 불과
- → 클라이언트는 HTML 내에 포함된 `<script>` 태그를 보고 서버에서 React 번들 파일(bundle.js)을 다운로드하여 실행. 이 번들 파일에는 `hydrateRoot` 함수가 포함 → 호출
- → `hydrateRoot` 함수가 호출되면 브라우저는 다음 작업을 수행
    - 서버에서 전달받은 정적 HTML과 React의 Virtual DOM을 비교하여 일치 여부 확인
    - HTML과 Virtual DOM이 일치하면, DOM 트리와 이벤트 핸들러를 React가 제어할 수 있도록 React DOM을 결합
    - 이 과정에서 서버에서 생성된 정적 HTML에 대해 `click` 등의 이벤트 핸들러가 추가되고, 컴포넌트의 `state`나 `props`에 따라 렌더링 진행 가능
    - 즉, 주도권이 리액트로 넘어온다.
